### PR TITLE
fix: Deprecate pause and resume in GameLoop

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

`GameLoop` methods `pause()` and `resume()` are marked as deprecated, with a recommendation to use `start()` and `stop()` instead. The reason they are deprecated is because their functionality is identical to that of the other pair of methods, so no reason to keep them around... Also, property `previous` was made final (with a deprecated getter for backward-compatibility), since changing it externally would break GameLoop's functionality.

Here's the summary of behavior changes:
- `start()` can now be called multiple times; previously calling `start()` twice would produce an error
- `stop()` now works the same way as `pause()`; previously calling `stop()` and then `start()` would produce negative times
- property `.previous` is no longer settable

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [x] No, this is *not* a breaking change.

## Related Issues

This is split out of #1234

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Comment by st-pasha** [comment]
Note: we may also make pause/resume the main methods, and redirect/deprecate start+stop.

File: packages/flame/lib/src/game/game_loop.dart [hunk 1]
```diff
@@ -2,47 +2,41 @@ import 'package:flutter/scheduler.dart';
...
   void stop() {
     _ticker.stop();
+    _previous = Duration.zero;
   }
 
   void dispose() {
     _ticker.dispose();
   }
 
-  void pause() {
```
Comment (luanpotter) [inline packages/flame/lib/src/game/game_loop.dart:35]: I don't think they were meant to be the same, IIRC @erickzanardo added those to fix a bug or something

Comment (erickzanardo) [reply packages/flame/lib/src/game/game_loop.dart:35]: Yes, I do remember something like this too, but I can remember exactly what but and why, we should hold this until we can remember

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: I also though we added these, but apparently these were added by someone called @erf 20 months ago, maybe they could explain?

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Here is the PR: https://github.com/flame-engine/flame/pull/302

Comment (erf) [reply packages/flame/lib/src/game/game_loop.dart:35]: @spydon hi, i took a look now at my PR where i simplified the game loop to use a Ticker. It looks like `start` and `stop` replaced the `scheduleTick` and `unscheduleTick` calls. The only different between those were that `previous` Duration were set to zero in `pause`. It think it would be fine to combine these calls. Not sure if there would be a case where e.g. the user would e.g. want to pause the callbacks, but still make the timer move forward  or something like that.

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Thanks for the input @erf!

Comment (erf) [reply packages/flame/lib/src/game/game_loop.dart:35]: No problem! I would perhaps also take a look at the `Ticker` inner workings, and see if you want to expose e.g. `isActive` or `hidden` to `GameLoop`.

Comment (st-pasha) [reply packages/flame/lib/src/game/game_loop.dart:35]: > The only different between those were that previous Duration were set to zero in pause.

Note that NOT setting duration to zero in stop, would have produced negative `dt` after the subsequent call to `start`, which is clearly a bug. The only reason it wasn't noticed is because `stop` was never used.

**Review by spydon — APPROVED** [review]
Lgtm, I don't think there ever was a reason to have both and everything seems to work fine without `pause` and `resume`.

File: packages/flame/lib/src/game/game_loop.dart (hunk 1 above)
Comment (luanpotter) [reply packages/flame/lib/src/game/game_loop.dart:35]: the bug was: whenever the widget was recreated, since we were using the engine state as the pause/unpause, it would restart the engine. we needed a difference between user manually calling pause (persistent) and engine being paused/unpaused by the framework.

can we test this new version under that scenario?

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: @luanpotter any suggestion for how this test should look?

Comment (luanpotter) [reply packages/flame/lib/src/game/game_loop.dart:35]: I didn't mean an automated test, just manually really!

Comment (st-pasha) [reply packages/flame/lib/src/game/game_loop.dart:35]: Interesting... So I tried the following manual test: launch examples/System/Pause-resume-engine, turn on a device frame, click on the helicopter to pause the engine, then click the button to rotate the device frame (which involves rebuilding the widget).

The helicopter starts spinning again, the engine doesn't stay paused. So I guess the test fails? I also tried the same test in the current `main`, and the outcome is the same: the engine restarts, test fails.

It looks like the reason is that the "ticker" of the engine belongs to the `GameLoop` class, which in turn belongs to the `GameRenderBox`, which in turns gets re-created via the `_GameRenderObjectWidget` whenever the `GameWidget` is rebuilt. So, when `GameWidget` is rebuilt, the `GameLoop` gets recreated, and it starts running.

In short, the bug that you describe currently exists in `main`, and this PR neither fixes it nor makes it worse. Should we perhaps open a separate issue for the bug?

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Yeah, then I think we can just as well merge this and file a bug for that behavior.

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Can you file an issue for that @st-pasha since you have the most context?

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Btw, in your test, did you declare the game outside of the `GameWidget`?

Comment (st-pasha) [reply packages/flame/lib/src/game/game_loop.dart:35]: I've tested with the standard example from our Dashbook, I'll have to check if it recreates the game instance upon frame rotation.

Comment (st-pasha) [reply packages/flame/lib/src/game/game_loop.dart:35]: Ok, I checked again -- after making sure that the game instance is declared only once in the examples app, the "bug" disappears (i.e. the game state keeps being paused after rotating the frame or changing window size, or opening the side menu). This happens both in the main branch and in this branch.

Comment (spydon) [reply packages/flame/lib/src/game/game_loop.dart:35]: Super, then it works as intended!
